src/gadget/monotone.rs

```diff
@@ -0,0 +1,259 @@
+use halo2::{
+    circuit::{Chip, Layouter},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, VirtualCells},
+    poly::Rotation,
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::{marker::PhantomData, u64};
+
+#[derive(Clone, Debug)]
+pub(crate) struct MonotoneConfig {
+    range_table: Column<Fixed>,
+}
+
+/// MonotoneChip helps to check if an advice column is monotone increasing
+/// within a range. With strict enabled, it disallows equality of two cell.
+pub(crate) struct MonotoneChip<F, const RANGE: usize, const INCR: bool, const STRICT: bool> {
+    config: MonotoneConfig,
+    _marker: PhantomData<F>,
+}
+
+pub(crate) type StrictMonoIncrChip<F, const RANGE: usize> = MonotoneChip<F, RANGE, true, true>;
+pub(crate) type NonStrictMonoIncrChip<F, const RANGE: usize> = MonotoneChip<F, RANGE, true, false>;
```

Comment:
I'm not sure that we need to define these here; we can just leave it up to the caller to instantiate `MonotoneChip` with the const generics they need.

Reply:
Fixed in second commit